be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/DbActor.scala

```diff
@@ -137,157 +181,249 @@ object DbActor extends AskPatternConstants {
+  // In the arguments, we might remove the initialChannelsMap and replace the channelNamesDb by the general Db??
+  // FIXME: find way to keep count of channel nb. Idea: store key "channelCount" with value int.getBytes to db and increment it with each channel created
+  sealed case class DbActor(mediatorRef: ActorRef, db: DB, DATABASE_FOLDER: String) extends Actor with ActorLogging {
+    
+    //helper function to determine if a channel should be created during write operations
+    private def writeCreateNewChannel(channel: Channel, message: Message): DbActorMessage = {
+      Try(db.get(channel.toString.getBytes)) match {
+        case Success(bytes) if bytes != null =>
+          log.info(s"Channel '$channel' already exists.")
+          DbActorAck()
+        case _ =>
+          val objectTypeOption = message.decodedData
+          val objectType = objectTypeOption match {
+            case Some(data) => data._object
+            case _ => ObjectType.LAO
+          }
+          // for now, we don't have meetup or roll call channels, so we just create Lao channels instead, easy to change if needed
+          val actualObjectType: ObjectType.ObjectType = if (objectType == ObjectType.ELECTION || objectType == ObjectType.CHIRP){
+            objectType
+          } else {
+            ObjectType.LAO
+          }
+          createChannel(channel, actualObjectType) match {
+            case DbActorAck() => DbActorAck()
+            case _ =>
+              log.info(s"Actor $self (db) encountered a problem while creating channel '$channel'")
+              DbActorNAck(ErrorCodes.SERVER_ERROR.id, s"Error when creating channel '$channel'")
+          }
+      }
     }
 
-    iterator.close()
-    DbActor(mediatorRef, initialChannelsMap.toMap, channelNamesDb, DATABASE_FOLDER, CHANNELS_FOLDER)
-  }
-
-  sealed case class DbActor(mediatorRef: ActorRef, initialChannelsMap: Map[Channel, DB], channelNamesDb: DB, DATABASE_FOLDER: String, CHANNELS_FOLDER: String) extends Actor with ActorLogging {
-    private val channelsMap: mutable.Map[Channel, DB] = initialChannelsMap.to(collection.mutable.Map)
+    //helper functions for the database key generation, could make them public later if needed elsewhere
+    private def generateMessageKey(channel: Channel, messageId: Hash): String = channel + "/" + messageId.toString
 
-    override def preStart(): Unit = {
-      log.info(
-        s"Actor $self (db) was initialised with a total of ${initialChannelsMap.size} recovered channels"
-      )
-      if (initialChannelsMap.size > DATABASE_MAX_CHANNELS) {
-        log.warning(
-          s"Actor $self (db) has surpassed a large number of active lao channels (${initialChannelsMap.size} > $DATABASE_MAX_CHANNELS)"
-        )
+    //may return null if the extractLaoId fails
+    private def generateLaoDataKey(channel: Channel): String = {
+      channel.decodeSubChannel match {
+        case Some(data) => new String(data, StandardCharsets.UTF_8) + "/data"
```

Comment:
What happens if we have a new channel in the format`laoId/data` ? 

Here, in `generateMessageKey`, in storing channels, in storing everything, I would suggest you use prefixes for all of your keys. Otherwise you might bump into key conflicts that will be extremely hard to debug.
It also makes it more obvious what types you're dealing with if (in the future) we need to look into the database to debug.

Reply:
This could be problematic indeed, however with our current implementation, we'll have to find some working prefix and incorporate it in the protocol. The documentation you were talking about in your general comment should be very useful for things like that. I will add it in the prefixes in Channel.scala

Reply:
I'm not sure Channel.scala is the best place. Prefixes would really be an internal implementation detail of the storage layer :)

Reply:
I suggested Channel.scala since the other prefixes were there but I will move it directly to DbActor since it effectively makes more sense